### DISCUSSION TITLE
How to interface jsbsim with an input on Python?

### DISCUSSION THREAD

**Discussion by WangKeji** [JSBSim-Team/jsbsim#418] [Q&A]
Greeting! I installed jsbsim and wanted to use Python to simulate the flight of an airplane . What I want to achieve is, given an input, including aileron, rudder, elevator and throttle, I can get its 12 statuses(u,v,w,p,q,r and so on) . It seems that FGFCS could achieve but I did not find FGFCS on Python module.
 ```
FGFCS * fcs = FDMExec->GetFCS();
 
fcs->SetDeCmd(ele_cmd_norm);
fcs->SetDePos(0, ele_cmd_rad);
fcs->SetDePos(1, ele_cmd_deg); 
fcs->SetDePos(2, ele_cmd__norm);
 
fcs->SetDaCmd(ail_cmd_norm); 
fcs->SetDaPos(0, ail_cmd_rad);
fcs->SetDaPos(1, ail_cmd_deg); 
fcs->SetDaPos(2, ail_cmd__norm);
```
So how can I achieve it？

**Comment by seanmcleod** [comment]
Yep, I'm not sure why FGFCS isn't exported.

https://github.com/JSBSim-Team/jsbsim/blob/af9532dd4b379b31b817ff943a92d7e3043c297b/python/jsbsim.pxd#L224-L233

However you can set the relevant control commands via the property tree that is exposed, e.g.

```python
import jsbsim

fdm = jsbsim.FGFDMExec('.', None)
fdm.load_script('scripts/c1721.xml')
fdm.run_ic()

while fdm.run():
  fdm['fcs/elevator-cmd-norm'] = elevatorCommand
  fdm['fcs/aileron-cmd-norm'] = aileronCommand
```

**Reply by bcoconni** [reply]
> Yep, I'm not sure why FGFCS isn't exported.

I guess nobody has ever needed to access `FGFCS` from Python until now.  This could be added quite easily if desperately needed. However, as @seanmcleod mentioned, the parameters you are trying to access can be read and written using Python via properties such as `fcs/elevator-cmd-norm` and its siblings.

**Reply by WangKeji** [reply]
Thanks a lot! I have another question now. While simulation I need to do something else at the same time. Which means, I want to use jsbsim to get airplane`s status within one step(right now I assume dt is 50ms). Then at this step, these status data need to be used to calculate something else(such as radar detection, missile hit etc.), output some key value, then continue the next step. Now I know how to initialize the initial state and simulate but this process is continuous, all I can get is an output .csv file. Maybe I can try to rewrite the initial file , restart the fdm at the begin of each step and read .csv file to get status data, but this seems too dumb☹.So is there some way to make jsbsim run as I want? I will be grateful for your kind help.

**Comment by seanmcleod** [comment]
So each call to `fdm.run()` executes a single simulation step, so simply insert your code in the loop, e.g.

```python
while fdm.run():
  # Get aircraft status
  lat = fdm['...']
  lon = fdm['...']
  altitude = fdm['....']
  speed = fdm['...']

  # Make use of aircraft status to perform calculations for radar detection, missile hit caclulation etc.
  performRadarDetection(lat, lon, altitude)
  performMissileHitDetection(lat, lon, altitude, speed)

  # Set control inputs for the next simulation time step
  fdm['fcs/elevator-cmd-norm'] = elevatorCommand
  fdm['fcs/aileron-cmd-norm'] = aileronCommand
```

**Reply by WangKeji** [reply]
My code right now is like
```
def step(airplane_command, ...):
    # Simulate aircraft flight according to commands in one step(0.05s)
    aircraft_status = aircraft_run(airplane_command, airplane_status_former_step)


    # Get status 
    lat = aircraft_status['...']
    lon = aircraft_status['...']
    altitude = aircraft_status['....']
    speed = aircraft_status['...']

    # Use status to perform calculation
    radar_detection = performRadarDetection(lat, lon, altitude)
    missile_hit = performMissileHitDetection(lat, lon, altitude, speed)

    # return some information
    return radar_detection, missile_hit ...
```
In actual operation, I can run function `step` to realize the simulation of the whole battlefield. So my question is how to merge fdm into `step` ?

**Comment by seanmcleod** [comment]
What about this?

```python

def step(airplane_command, fdm):
    # Simulate aircraft flight according to commands in one step(0.05s)
    fdm['fcs/elevator-cmd-norm'] = airplane_command.elevatorCommand
    fdm['fcs/aileron-cmd-norm'] = airplane_command.aileronCommand
    ....
    fdm.run()

    # Get status 
    lat = fdm['...']
    lon = fdm['...']
    altitude = fdm['....']
    speed = fdm['...']

    # Use status to perform calculation
    radar_detection = performRadarDetection(lat, lon, altitude)
    missile_hit = performMissileHitDetection(lat, lon, altitude, speed)

    # return some information
    return radar_detection, missile_hit ...
```

**Reply by WangKeji** [reply]
Copy that, I'll try to see if it works. Really grateful for your help😊. Have a nice day.
